### DISCUSSION TITLE
User-friendly tooling to localize / translate the user interface in different languages

### DISCUSSION THREAD

**Discussion by horus68** [edrlab/thorium-reader#1779] [Developer corner]
Is there a collaborative site to translate the software interface (see Crowdin https://crowdin.com/ or Transifex https://www.transifex.com/)?
Would love to collaborate with translations but i prefer to use a translation platform when translating!

**Comment by danielweck** [comment]
This makes sense, thank you for your feedback!

The localization files use a simple JSON format with nested unique identifiers / hierarchical labels:
https://github.com/edrlab/thorium-reader/blob/develop/src/resources/locales/en.json
Is that compatible with Crowin and/or Transifex?

**Comment by horus68** [comment]
I tested it and it works perfectly fine using Transifex or Crowdin
Both have free plans for open source software (upon request).
I use both, but I do prefere Crowdin (translator or management). Its up to you!

If using Transifex upload the first file using format Key-value JSON
![Resources](https://user-images.githubusercontent.com/3055443/86940963-47cbef00-c13b-11ea-97db-3d917a92289f.png)

Crowdin detects the format automatically (just use UTF-8 source files when uuploading.

For visuals...
- Your file under Transifex 
![transifex](https://user-images.githubusercontent.com/3055443/86941025-5914fb80-c13b-11ea-8e90-0e1699c667ad.png)

- Your file under Crowdin
![crowdin](https://user-images.githubusercontent.com/3055443/86941053-629e6380-c13b-11ea-9978-ebdd512e9b7d.png)

**Comment by danielweck** [comment]
Thank you for testing this for us. Very helpful.

**Comment by horus68** [comment]
Any news on this?

**Comment by danielweck** [comment]
No, sorry (other development priorities at the moment).
Pull Requests welcome :)

**Comment by horus68** [comment]
It's not a question of create a pull request, I've already done what can be done by a regular user (testing the platforms usage).
Now the task has to be done by someone related with the software project management (using an oficial email account to request an opensource plan)
Note: the email account used to start the project will be the "owner" and will have additional permissions on using API and CLI tools. The owner can then add administrators to help manage the project.

**1- Take a decision on the platform to use (Transifex or Crowdin)**
As the platforms are almost equivalents for users the decision will be mostly on the Dev "feeling" and the way each CLI/API tools are used to automate the release processes (upload master language and download translations to integrate with Github). You need to test and decide!
Note: updates and downloads can also be done by hand by using each platform web interface but maybe later you can prefer to automate things:
Crowdin API: https://support.crowdin.com/api/api-integration-setup/
Transifex CLI: https://docs.transifex.com/client/introduction
Transifex and Github: https://docs.transifex.com/transifex-github-integrations/github-tx-ui

**2- Create a translation project to hold languages on the chosen platform**
Crowdin: Open-source project setup request form - https://crowdin.com/page/open-source-project-setup-request
Transifex: https://docs.transifex.com/projects/open-source-project

**3- Open target languages**
and invite translators and proofreaders for each language (and regular maintenance)


***Where can I help?***
Step 1: nothing for me here, sorry!
Step 2: I can help you after a project has been created, you need to invite me as an administrator for the translation project)
Step 3: I can take that task pro bono (I already do that for some other software). I will step down when asked!

So.... open a project with Crowdin, invite me and let it roll!

To invite me you can use the email pauloizidoro@ gmail or my profile name for each platform:
Crowdin: https://crowdin.com/profile/horus68
Transifex: https://www.transifex.com/user/profile/horus68/

**Comment by danielweck** [comment]
Thank you for detailing the required actions, this is really useful.

**Comment by llemeurfr** [comment]
We never had the time to tackle this issue so far, it is now a good time to do it.

So we'll start by testing and choosing between **Transifex**, **Crowdin** and **Weblate**, the latter being recommended by another partner.

**Comment by horus68** [comment]
So... are we there yet?
Keeping translation updated has been an ugly task and time consuming! 
As there is no changes since so long, just don't try to make this a perfect thing and let translatores move on. If in a year or two you find a better way just change it then to a new system. Translators need to translate and stop reading merged PR and actions in gitub!

So, my 10 steps plan is (this will take 10 minutes for you, and for me it will be 2 hours work):
1- Take a decision on the platform to use (Transifex or Crowdin) I would prefer Crowdin
2- Create an empty translation project
3- Make me an admin for the project (on transifex and Crowdin my user: horus68)
4- I will setup project and upload all translations from Github
5- I will send invitations for translators and proofreaders for each language (based on their Github work)
6- I will step down from admin and focus on my language team (If you want you can keep me as "project manager" so I can update the master language file)
7- Translators will translate their language
8- Proofreaders will proofread the translation for their language
9- Sync master language files into Crowdin: You will have to upload the master language files to Crowdin when it is updated in Github.
10- Sync translations into Github - Before each software release you have to download/sync translated files and merge it to Github

**Comment by llemeurfr** [comment]
Hi @horus68, I'll put the subject on the table in the scope of the current evolution of the UI or Thorium, which will involved some evolution of localization strings.

**Comment by horus68** [comment]
Are we there yet?!
The new version really needs a tool for translations!

**Reply by danielweck** [reply]
Hello, we will very likely use GitLocalize as this free service has met many of our requirements during our tests, compared with other tools. We will let you know when this is ready (soon, I am cleaning up the en-US "master" file)

**Reply by danielweck** [reply]
Hello @horus68 could you please help us test the GitLocalize instance, for example:

https://gitlocalize.com/repo/9626/pt-br/src/resources/locales/en.json

https://gitlocalize.com/repo/9626/pt-PT/src/resources/locales/en.json

**Reply by danielweck** [reply]
note that we are nearly done with the en-US base/canonical reference, so please do not start translating the missing items just yet, this is just to trial the system

**Reply by danielweck** [reply]
I will add you as a "translator" as soon as you are registered with your GitHub account in GitLocalize

<img width="546" alt="Screenshot 2024-05-15 at 12 28 29" src="https://github.com/edrlab/thorium-reader/assets/625201/1a976a28-791d-475e-a62b-e9ecd2117da2">

**Comment by horus68** [comment]
> Hello @horus68 could you please help us test the GitLocalize instance

Done!
https://gitlocalize.com/repo/9626/pt-PT/review/71335

**Comment by gautierchomel** [comment]
We want to reopen this discussion because of the ongoing issues with Gitlocalize. 

Like @horus68 my choice would be crowdin.

**Comment by gautierchomel** [comment]
Developer preference went to Weblate :) 
https://hosted.weblate.org/projects/thorium-reader
Please let us know if you encounter difficulties and would like us to maintain gitlocalize aside, otherwise we'll stop using gitlocalize.
